### PR TITLE
[cudapoa] adding new tests for different banding modes in NW

### LINKED ISSUE #584: [cudapoa] add tests for different band modes
there are different variations of NW algorithm (full-band, static-band, adaptive-band, traceback static and traceback adaptive bands). We need to add some tests for each case.

### PR DESCRIPTION
**Pull request by r-mafi**
closes #584
added 4 new tests to validate results of static band, adaptive band, static band with traceback and adaptive band with traceback against full-band NW.

### CODE REVIEW

File: cudapoa/src/cudapoa_nw_tb_banded.cuh [hunk 1]
```diff
@@ -642,6 +642,134 @@ __device__ __forceinline__
     return aligned_nodes;
 }
 
+// global kernel used in testing, hence uses int16_t for SizeT and ScoreT,
+// may need to change if test inputs change to long reads
+template <typename SizeT>
```
Comment (ahehn-nv) [inline cudapoa/src/cudapoa_nw_tb_banded.cuh:647]: See other comments. template needed? maybe dispatch on host?

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 2]
```diff
@@ -304,6 +305,220 @@ TEST_P(NWTest, TestNWCorrectness)
...
+    int32_t mismatch_score;
+    int32_t match_score;
+    SizeT* aligned_nodes; //local; to store num of nodes aligned (length of alignment_graph and alignment_read)
+    BandMode band_mode = traceback ? (adaptive ? BandMode::adaptive_band_traceback : BandMode::static_band_traceback)
+                                   : (adaptive ? BandMode::adaptive_band : BandMode::static_band);
+    BatchConfig batch_size(1024 /*max_sequence_size*/, 2 /*max_sequences_per_poa*/,
+                           128 /*= band_width*/, band_mode);
+
+    //allocate unified memory so they can be accessed by both host and device.
+    GW_CU_CHECK_ERR(cudaMallocManaged((void**)&nodes, batch_size.max_nodes_per_graph * sizeof(uint8_t)));
```
Comment (ahehn-nv) [inline cudapoa/tests/Test_CudapoaNW.cu:340]: Optional: Are the explicit casts to `(void**)` really necessary?

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 3]
```diff
@@ -304,6 +305,220 @@ TEST_P(NWTest, TestNWCorrectness)
...
+        GW_CU_CHECK_ERR(cudaFree(traces));
+    }
+
+    return res;
+}
+
+class NWbandedTest : public ::testing::Test
+{
+public:
+    BasicNW* nw;
```
Comment (ahehn-nv) [inline cudapoa/tests/Test_CudapoaNW.cu:457]: Could you use a `std::unique_ptr<BasicNW>` here? Then you should be able to drop the `TearDown()` function.

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 4]
```diff
@@ -304,6 +305,220 @@ TEST_P(NWTest, TestNWCorrectness)
...
+    else
+    {
+        GW_CU_CHECK_ERR(cudaMallocManaged((void**)&scores, batch_size.max_nodes_per_graph * batch_size.matrix_sequence_dimension * sizeof(int16_t)));
+    }
+
+    //initialize all 'count' buffers
+    memset((void**)incoming_edge_count, 0, batch_size.max_nodes_per_graph * sizeof(uint16_t));
+    memset((void**)outgoing_edge_count, 0, batch_size.max_nodes_per_graph * sizeof(uint16_t));
+    memset((void**)node_id_to_pos, 0, batch_size.max_nodes_per_graph * sizeof(SizeT));
+    memset((void**)scores, 0, batch_size.max_nodes_per_graph * batch_size.matrix_sequence_dimension * sizeof(int16_t));
```
Comment (ahehn-nv) [inline cudapoa/tests/Test_CudapoaNW.cu:365]: If any casts are needed, those should be casts to `(void*)`.

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 5]
```diff
@@ -304,6 +305,220 @@ TEST_P(NWTest, TestNWCorrectness)
...
+    uint8_t* read;
+    uint16_t read_count; //local
+    int16_t* scores;
+    int16_t* traces;
+    SizeT* alignment_graph;
+    SizeT* alignment_read;
+    int32_t gap_score;
+    int32_t mismatch_score;
+    int32_t match_score;
+    SizeT* aligned_nodes; //local; to store num of nodes aligned (length of alignment_graph and alignment_read)
```
Comment (ahehn-nv) [inline cudapoa/tests/Test_CudapoaNW.cu:333]: Initialize the pointers to `nullptr` to fail hard in case you forget to reset any of them. :boom: ;)

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 6]
```diff
@@ -304,6 +305,220 @@ TEST_P(NWTest, TestNWCorrectness)
...
+
+    //calculate edge counts on host
+    obj.get_graph_buffers(incoming_edges, incoming_edge_count,
+                          outgoing_edges, outgoing_edge_count,
+                          nodes, &graph_count,
+                          graph, node_id_to_pos);
+    obj.get_read_buffers(read, &read_count);
+    gap_score      = BasicNW::gap_score_;
+    mismatch_score = BasicNW::mismatch_score_;
+    match_score    = BasicNW::match_score_;
```
Comment (ahehn-nv) [inline cudapoa/tests/Test_CudapoaNW.cu:375]: Why don't you declare the `*_score` variables here and initialize them on the spot?

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 5 above)
Comment (ahehn-nv) [reply cudapoa/tests/Test_CudapoaNW.cu:333]: Also initialize `graph_count` and `read_count` to a sensible value? Probably 0?

**Review by ahehn-nv — CHANGES_REQUESTED** [review]
Looks good overall. I'd like to propose some minor changes.

File: cudapoa/src/cudapoa_nw.cuh [hunk 7]
```diff
@@ -472,22 +474,22 @@ __global__ void runNeedlemanWunschKernel(uint8_t* nodes,
                                          int32_t match_score,
                                          SizeT* aligned_nodes)
 {
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw.cuh:476]: here I copy a summary of my discussion with Andreas on slack:
removing `SizeT` template parameter results in linking error of multiple definitions found. Adding `inline` fixes the problem, but adds multiple instances of this `warning: inline qualifier ignored for "__global__"`. Another option is to use `static` instead of `inline`, where it creates a separate copy of the kernel per object file. 
To avoid compiler warnings, for now the kernel is kept templetized and `static_assert` is added to verify `SizeT` is `int16_t`

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 8]
```diff
@@ -556,6 +556,122 @@ __device__ __forceinline__
     return aligned_nodes;
 }
 
+// global kernel used in testing, hence uses int16_t for SizeT and ScoreT,
+// may need to change if test inputs change to long reads
+template <typename SizeT>
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:561]: removed `SizeT` instances in tests files, replaced with `int16_t`. But since `runNW` is defined in `cudapoa_nw.cuh`, to avoid linking errors, it's being kept templetized.

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 9]
```diff
@@ -556,6 +556,122 @@ __device__ __forceinline__
...
+                                               int32_t band_width,
+                                               int32_t gap_score,
+                                               int32_t mismatch_score,
+                                               int32_t match_score,
+                                               SizeT* aligned_nodes,
+                                               bool adaptive)
+{
+    float banded_buffer_size = static_cast<float>(max_nodes_per_graph) * static_cast<float>(scores_width);
+
+    if (adaptive)
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:585]: since this is only for test case use with a small dataset, it does not really impact on performance. Changed it not to pass `bool adaptive` to device kernel.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 2 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:340]: good point, unnecessary, removed.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 4 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:365]: cast was not needed, removed them all.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 3 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:457]: done.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 5 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:333]: done.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 6 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:375]: done.

File: cudapoa/src/cudapoa_nw_tb_banded.cuh (hunk 1 above)
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_tb_banded.cuh:647]: addressed above.

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 10]
```diff
@@ -185,86 +186,81 @@ std::vector<NWTestPair> getNWTestCases()
...
+          match_score,
+          aligned_nodes);
 
     GW_CU_CHECK_ERR(cudaDeviceSynchronize());
 
     //input and output buffers are the same ones in unified memory, so the results are updated in place
     //results are stored in alignment_graph and alignment_read; return string representation of those
-    auto res = std::make_pair(genomeworks::stringutils::array_to_string<SizeT>(alignment_graph, *aligned_nodes, ","),
-                              genomeworks::stringutils::array_to_string<SizeT>(alignment_read, *aligned_nodes, ","));
+    auto res = std::make_pair(genomeworks::stringutils::array_to_string<int16_t>(alignment_graph, *aligned_nodes, ","),
```
Comment (ahehn-nv) [inline cudapoa/tests/Test_CudapoaNW.cu:262]: Optional: template argument `<int16_t>` not needed - let the compiler deduce it automatically from the `alignment_graph` argument.

**Review by ahehn-nv — CHANGES_REQUESTED** [review]
I think it's mostly some missing static_asserts at the right place. Otherwise suggestions for mostly cosmetic changes.

File: cudapoa/src/cudapoa_generate_consensus.cuh [hunk 11]
```diff
@@ -410,6 +410,8 @@ void generateConsensusTestHost(uint8_t* nodes,
                                uint16_t* node_alignment_count,
                                int32_t max_limit_consensus_size)
 {
+    static_assert(std::is_same<SizeT, int16_t>::value, "This function only accepts int16_t as SizeT.");
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_generate_consensus.cuh:413]: that's right, got tricked by the symmetry of the code, but it is not similar to NW cases where it helps to add the static_assert. removed it from here.

File: cudapoa/src/cudapoa_nw.cuh [hunk 12]
```diff
@@ -510,23 +512,25 @@ void runNW(uint8_t* nodes,
...
-                                        read_length,
-                                        scores,
-                                        scores_width,
-                                        alignment_graph,
-                                        alignment_read,
-                                        gap_score,
-                                        mismatch_score,
-                                        match_score,
-                                        aligned_nodes);
+    static_assert(std::is_same<SizeT, int16_t>::value, "This function only accepts int16_t as SizeT.");
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw.cuh:515]: `runNW()` calls `runNeedlemanWunschKernel()` where in the latter kernel, we assume `SizeT == int16_t` and there is a comment clarifying this assumption. I agree it makes more sense to move the assert right to `runNeedlemanWunschKernel()`. done.

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 13]
```diff
@@ -556,6 +556,124 @@ __device__ __forceinline__
...
+                                               int32_t scores_width,
+                                               int32_t max_nodes_per_graph,
+                                               SizeT* alignment_graph,
+                                               SizeT* alignment_read,
+                                               int32_t band_width,
+                                               int32_t gap_score,
+                                               int32_t mismatch_score,
+                                               int32_t match_score,
+                                               SizeT* aligned_nodes)
+{
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:581]: done.

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 14]
```diff
@@ -556,6 +556,124 @@ __device__ __forceinline__
...
+                 SizeT* alignment_graph,
+                 SizeT* alignment_read,
+                 int32_t band_width,
+                 int32_t gap_score,
+                 int32_t mismatch_score,
+                 int32_t match_score,
+                 SizeT* aligned_nodes,
+                 bool adaptive)
+{
+    static_assert(std::is_same<SizeT, int16_t>::value, "This function only accepts int16_t as SizeT.");
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:627]: done.

File: cudapoa/src/cudapoa_nw_tb_banded.cuh [hunk 15]
```diff
@@ -642,6 +642,136 @@ __device__ __forceinline__
...
+                   SizeT* alignment_read,
+                   int32_t band_width,
+                   int32_t score_matrix_height,
+                   int32_t gap_score,
+                   int32_t mismatch_score,
+                   int32_t match_score,
+                   SizeT* aligned_nodes,
+                   bool adaptive)
+{
+    static_assert(std::is_same<SizeT, int16_t>::value, "This function only accepts int16_t as SizeT.");
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_tb_banded.cuh:748]: done.

File: cudapoa/src/cudapoa_topsort.cuh [hunk 16]
```diff
@@ -225,6 +225,8 @@ void runTopSort(SizeT* sorted_poa,
                 uint16_t* outgoing_edge_count,
                 uint16_t* local_incoming_edge_count)
 {
+    static_assert(std::is_same<SizeT, int16_t>::value, "This function only accepts int16_t as SizeT.");
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_topsort.cuh:228]: done.

File: cudapoa/tests/Test_CudapoaGenerateConsensus.cu [hunk 17]
```diff
@@ -169,20 +169,20 @@ std::string testGenerateConsensus(const BasicGenerateConsensus& obj)
 {
     //declare device buffer
     uint8_t* nodes;
-    SizeT* node_count;
-    SizeT* graph;
-    SizeT* node_id_to_pos;
-    SizeT* incoming_edges;
+    int16_t* node_count;
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaGenerateConsensus.cu:172]: sure, done.

File: cudapoa/tests/Test_CudapoaAddAlignment.cu [hunk 18]
```diff
@@ -232,58 +232,57 @@ std::vector<AddAlginmentTestPair> getAddAlignmentTestCases()
 // host function for calling the kernel to test topsort device function.
 BasicGraph testAddAlignment(const BasicAlignment& obj)
 {
-    typedef int16_t SizeT;
     //declare device buffer
     uint8_t* nodes;
-    SizeT* node_count;
-    SizeT* node_alignments;
+    int16_t* node_count;
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaAddAlignment.cu:237]: done, also fixed it in `Test_CudapoaTopSort.cpp`

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 10 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:262]: right, that's result of using replace all! removed.

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 19]
```diff
@@ -556,6 +556,124 @@ __device__ __forceinline__
...
+                                               SizeT* alignment_read,
+                                               int32_t band_width,
+                                               int32_t gap_score,
+                                               int32_t mismatch_score,
+                                               int32_t match_score,
+                                               SizeT* aligned_nodes)
+{
+    float banded_buffer_size = static_cast<float>(max_nodes_per_graph) * static_cast<float>(scores_width);
+
+    *aligned_nodes = runNeedlemanWunschBanded<uint8_t, int16_t, int16_t, ADAPTIVE>(nodes,
```
Comment (atadkase) [inline cudapoa/src/cudapoa_nw_banded.cuh:584]: nit: Can this function be called something on the lines of `needlemanWunschBanded`? The top level function is also called `runNWbanded`

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 20]
```diff
@@ -556,6 +556,124 @@ __device__ __forceinline__
...
+                                               int32_t max_nodes_per_graph,
+                                               SizeT* alignment_graph,
+                                               SizeT* alignment_read,
+                                               int32_t band_width,
+                                               int32_t gap_score,
+                                               int32_t mismatch_score,
+                                               int32_t match_score,
+                                               SizeT* aligned_nodes)
+{
+    float banded_buffer_size = static_cast<float>(max_nodes_per_graph) * static_cast<float>(scores_width);
```
Comment (atadkase) [inline cudapoa/src/cudapoa_nw_banded.cuh:582]: Will precision round-off be an issue for a corner case? (I realize that it might be pretty rare)

File: cudapoa/src/cudapoa_nw_tb_banded.cuh [hunk 21]
```diff
@@ -642,6 +642,136 @@ __device__ __forceinline__
...
+                                                 int32_t mismatch_score,
+                                                 int32_t match_score,
+                                                 SizeT* aligned_nodes,
+                                                 bool adaptive)
+{
+    float banded_buffer_size = static_cast<float>(max_nodes_per_graph) * static_cast<float>(scores_width);
+
+    if (adaptive)
+    {
+        *aligned_nodes = runNeedlemanWunschBandedTraceback<uint8_t, int16_t,
```
Comment (atadkase) [inline cudapoa/src/cudapoa_nw_tb_banded.cuh:675]: nit: same renaming suggestion as above.

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 22]
```diff
@@ -304,6 +300,213 @@ TEST_P(NWTest, TestNWCorrectness)
...
+        {
+            outgoing_edges[i].push_back(i + 1);
+        }
+        std::vector<uint8_t> read(read_str.begin(), read_str.end());
+        // setup nw
+        nw = std::make_unique<BasicNW>(nodes, sorted_graph, outgoing_edges, read);
+    }
+};
+
+TEST_F(NWbandedTest, NWSaticBandvsFull)
```
Comment (atadkase) [inline cudapoa/tests/Test_CudapoaNW.cu:470]:
```suggestion
TEST_F(NWbandedTest, NWStaticBandvsFull)
```

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 23]
```diff
@@ -304,6 +300,213 @@ TEST_P(NWTest, TestNWCorrectness)
...
+class NWbandedTest : public ::testing::Test
+{
+public:
+    std::unique_ptr<BasicNW> nw;
+
+public:
+    void SetUp()
+    {
+        // initialize nw graph and read with the following data
+        std::string nodes_str = "TTTAACCTAATAAATCAGTGAAGATTTAAAATATGATAATTATTGATTTTGGTGAGAGTGCAAAGAAATTTGTTACCCTCATAAGCTGAGCAGACAGATAAGATAGAAAAACAGAAGATAGAATATTAAAACCATGATAGGTACAGACTGAAAAATTCTTGGATAAATATTAAAATTTAGGCTTTAGTAGTAGATTGATGACTGTGAGGAAAAAGGATGTCCAATTGTTGAGTGACATGTAGAATGCCTTAAAATAATTTTACACGTCACTGAAAGCTATATTTATATTCAGGAAGGATATATCCCAGTCATGATTTTCTTAATAAGTTGCCCCATTTTCCAAGTTTAGCTAATTAACATTTATGTCTTCTATAATCAGGAATAGTCATTAACTGACACAGAAACAATTGGAAGCATATGTAGCCAAAAACATAAAAATTATTGCATCCAAATAATGATAAAGTAAAATATTAAAAAATATAGTCTTCTAAAT";
```
Comment (atadkase) [inline cudapoa/tests/Test_CudapoaNW.cu:453]: Optional: I currently see this as a single test case for this suite. Will it be possible to modularize it out so that we can add additional test cases for corner cases we see later?

File: cudapoa/tests/Test_CudapoaNW.cu [hunk 24]
```diff
@@ -304,6 +300,213 @@ TEST_P(NWTest, TestNWCorrectness)
...
+{
+    auto full_alignment_results  = testNW(*nw);
+    auto adaptive_banded_results = testNWbanded(*nw, true);
+    // verify alignment_graph
+    EXPECT_EQ(full_alignment_results.first, adaptive_banded_results.first);
+    // verify alignment_read
+    EXPECT_EQ(full_alignment_results.second, adaptive_banded_results.second);
+}
+
+TEST_F(NWbandedTest, NWSaticBandTracebackvsFull)
```
Comment (atadkase) [inline cudapoa/tests/Test_CudapoaNW.cu:490]:
```suggestion
TEST_F(NWbandedTest, NWStaticBandTracebackvsFull)
```

**Review by atadkase — CHANGES_REQUESTED** [review]
LGTM. Minor cosmetic changes w.r.t. renaming and typos. Also a possible re-org for test design.

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 25]
```diff
@@ -556,6 +556,124 @@ __device__ __forceinline__
     return aligned_nodes;
 }
 
+// global kernel used in testing, hence uses int16_t for SizeT and ScoreT,
+// may need to change if test inputs change to long reads
+template <typename SizeT, bool ADAPTIVE>
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:561]: done. I checked some other similar cases, found `TRACEBACK`, `BM`, `MSA` as template parameters in `generatePOAKernel()`. Changed the first, but left the other two as it is.

File: cudapoa/src/cudapoa_nw_tb_banded.cuh [hunk 26]
```diff
@@ -642,6 +642,136 @@ __device__ __forceinline__
...
+                                                 SizeT* alignment_graph,
+                                                 SizeT* alignment_read,
+                                                 int32_t band_width,
+                                                 int32_t score_matrix_height,
+                                                 int32_t gap_score,
+                                                 int32_t mismatch_score,
+                                                 int32_t match_score,
+                                                 SizeT* aligned_nodes,
+                                                 bool adaptive)
+{
```
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_tb_banded.cuh:670]: done.

File: cudapoa/src/cudapoa_nw_banded.cuh (hunk 20 above)
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:582]: `banded_buffer_size` is to determine maximum capacity for scores buffer (or in some similar cases, traceback buffer). This maximum is based on band-width and maximum graph size, which the latter is selected heuristically, e.g. x3 of max sequence size. So overall what `banded_buffer_size` implies can certainly tolerate some error. The reason behind strange-looking cast to float is minimizing register count, somehow compiler did not like int64_t and switching to float would save around 4-5 registers if I remember correctly.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 23 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:453]: This is implemented as `TEST_F`, test with fixtures, which allow multiple tests use the same initialization which takes place in `SetUp()`. This was suitable for this case where we have 4 tests for validating 4 different band modes against full-alignment. All 4 using the same setup.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 22 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:470]: done.

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 24 above)
Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:490]: done.

File: cudapoa/src/cudapoa_nw_banded.cuh (hunk 25 above)
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:561]: also found `CPT` as template param, left unchanged.

File: cudapoa/src/cudapoa_nw_banded.cuh (hunk 19 above)
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:584]: changed `runNeedlemanWunsch*()` device kernels to `needlemanWunch*()`, hopefully that won't upset Mr. Saul Ben Needleman!

File: cudapoa/src/cudapoa_nw_tb_banded.cuh (hunk 21 above)
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_tb_banded.cuh:675]: changed here as well as in `cudapoa_nw.cuh`

File: cudapoa/tests/Test_CudapoaNW.cu (hunk 23 above)
Comment (atadkase) [reply cudapoa/tests/Test_CudapoaNW.cu:453]: I understand. I was recommending parameterizing the two strings used to create the graph.

Comment (r-mafi) [reply cudapoa/tests/Test_CudapoaNW.cu:453]: This is an extension to the previous NW test which used only a very short string (4-5 bases). In the even of designing new tests for NW that could use this setup, they can be added as new `TEST_F` cases, otherwise they can use e.g. `sample_windows.txt`. Therefore I'd rather leave it unchanged.

File: cudapoa/src/cudapoa_nw_banded.cuh (hunk 20 above)
Comment (atadkase) [reply cudapoa/src/cudapoa_nw_banded.cuh:584]: Ok!

**Review by atadkase — APPROVED** [review]
Approved from my side. Let's wait to hear back from Andreas.

File: cudapoa/src/cudapoa_nw.cuh (hunk 7 above)
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw.cuh:476]: done.

File: cudapoa/src/cudapoa_topsort.cuh (hunk 16 above)
[the remark this answers is not in the document]
Comment (ahehn-nv) [reply cudapoa/src/cudapoa_topsort.cuh:228]: Is there a commit missing?

**Review by ahehn-nv — APPROVED** [review]
There's one unresolved optional item. Maybe you forgot to commit or push? Otherwise everything is good now. Approved.

File: cudapoa/src/cudapoa_topsort.cuh (hunk 16 above)
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_topsort.cuh:228]: strange, not sure what happened here, fixed.

**Review by ahehn-nv — APPROVED** [review]

**Comment by r-mafi** [comment]
rerun tests
